zebra-consensus/src/block/check.rs

```diff
@@ -51,6 +57,38 @@ pub fn is_equihash_solution_valid(header: &Header) -> Result<(), equihash::Error
 /// accepted." [§7.5][7.5]
 ///
 /// [7.5]: https://zips.z.cash/protocol/protocol.pdf#blockheader
-pub fn is_time_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
+pub fn time_is_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
     header.is_time_valid_at(now)
 }
+
+/// [3.9]: https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts
+pub fn subsidy_is_correct(network: Network, block: &Block) -> Result<(), BoxError> {
+    let height = block
+        .coinbase_height()
+        .expect("always called on blocks with a coinbase height");
+
+    let coinbase = block.transactions.get(0).ok_or("no coinbase transaction")?;
+    let outputs = coinbase.outputs();
+
+    let canopy_height = Canopy.activation_height(network).ok_or("no canopy")?;
+    if height >= canopy_height {
+        panic!("Can't validate Canopy yet");
+    }
+
+    // validate founders reward and miner subsidy
+    if height > Height(0) && height <= Height(Params::LAST_FOUNDER_REWARD_HEIGHT) {
+        let block_subsidy = subsidies::block_subsidy(height, Network::Mainnet) as i64;
+        let miner_subsidy = subsidies::miner_subsidy(height, Network::Mainnet) as i64;
+        let values = || {
+            outputs
+                .iter()
+                .map(|o| o.value.try_into().expect("value will be a valid amount"))
+        };
+        if values().any(|value: i64| value == block_subsidy / 5 && value != miner_subsidy)
+            && values().any(|value| value == miner_subsidy && value != block_subsidy / 5)
+        {
+            return Ok(());
+        }
+    }
+    Err("subsidy is not correct")?
```

Comment:
Updated error at https://github.com/ZcashFoundation/zebra/pull/1051/commits/fd5be036023094b184da24e25e8ade0cf6838b1b however this will be changed after https://github.com/ZcashFoundation/zebra/issues/1029